### ISSUE TITLE
Llama perf test terminates with std::out_of_range

### ISSUE THREAD

**Issue by cglagovichTT** [tenstorrent/tt-metal#8930]
https://github.com/tenstorrent/tt-metal/actions/runs/9235290032/job/25409990194

Perf pipeline on main gives this indexing error
```
terminate called after throwing an instance of 'std::out_of_range'
  what():  _Map_base::at
```

**Comment by cglagovichTT** [comment]
```
Thread 0x00007f3e43b24740 (most recent call first):
  File "/home/cglagovich/tt-metal/models/experimental/llama2_70b/tt/llama_decoder_optimized.py", line 305 in decode_forward
  File "/home/cglagovich/tt-metal/models/experimental/llama2_70b/tt/llama_decoder_optimized.py", line 279 in __call__
```

**Comment by cglagovichTT** [comment]
Error appears with 20 layers but not 10 or fewer

**Comment by cglagovichTT** [comment]
```
terminate called after throwing an instance of 'std::out_of_range'
  what():  _Map_base::at
--Type <RET> for more, q to quit, c to continue without paging--

Thread 217 "python" received signal SIGABRT, Aborted.
[Switching to Thread 0x7ffd427fc700 (LWP 3426703)]
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7dec859 in __GI_abort () at abort.c:79
#2  0x00007ffff45df8d1 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ffff45eb37c in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff45eb3e7 in std::terminate() () from /lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff45eb699 in __cxa_throw () from /lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007ffff45e233c in std::__throw_out_of_range(char const*) () from /lib/x86_64-linux-gnu/libstdc++.so.6
#7  0x00007fff88e2d66a in std::_Function_handler<void (tt::tt_metal::Device*), tt::tt_metal::Tensor::deallocate(bool)::$_0::operator()<tt::tt_metal::MultiDeviceStorage>(tt::tt_metal::MultiDeviceStorage&) const::{lambda(tt::tt_metal::Device*)#1}>::_M_invoke(std::_Any_data const&, tt::tt_metal::Device*&&) () from /home/cglagovich/tt-metal/build/lib/libtt_eager.so
#8  0x00007fff88e2d7cc in std::_Function_handler<void (), tt::tt_metal::Tensor::deallocate(bool)::$_0::operator()<tt::tt_metal::MultiDeviceStorage>(tt::tt_metal::MultiDeviceStorage&) const::{lambda()#1}>::_M_invoke(std::_Any_data const&) ()
   from /home/cglagovich/tt-metal/build/lib/libtt_eager.so
#9  0x00007fff88874f32 in tt::WorkExecutor::run_worker() () from /home/cglagovich/tt-metal/build/lib/libtt_metal.so
#10 0x00007ffff4617df4 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#11 0x00007ffff7daf609 in start_thread (arg=<optimized out>) at pthread_create.c:477
```

**Comment by cglagovichTT** [comment]
Another place I see the error from:
```
terminate called after throwing an instance of 'std::out_of_range'
  what():  _Map_base::at
Fatal Python error: Aborted

Thread 0x00007f7181ff7700 (most recent call first):
  File "/usr/lib/python3.8/threading.py", line 306 in wait
  File "/usr/lib/python3.8/threading.py", line 558 in wait
  File "/home/cglagovich/tt-metal/python_env/lib/python3.8/site-packages/tqdm/_monitor.py", line 60 in run
  File "/usr/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/usr/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007f73861fc740 (most recent call first):
  File "/home/cglagovich/tt-metal/ttnn/ttnn/core.py", line 37 in is_tensor_storage_on_device
  File "/home/cglagovich/tt-metal/ttnn/ttnn/operations/core.py", line 352 in to_torch
  ```

**Comment by cglagovichTT** [comment]
```
                 Always | FATAL    | Worker ID out of bounds 2 after deallocate
terminate called after throwing an instance of 'std::runtime_error'
  what():  TT_FATAL @ ../tt_eager/tensor/tensor.cpp:195: s.buffers.find(worker->id()) != s.buffers.end()
info:
Worker ID out of bounds 2 after deallocate
backtrace:
```

**Comment by cglagovichTT** [comment]
```
                 Always | FATAL    | Worker ID out of bounds 7 after deallocate
terminate called after throwing an instance of 'std::runtime_error'
  what():  TT_FATAL @ ../tt_eager/tensor/tensor.cpp:195: s.buffers.find(worker->id()) != s.buffers.end()
info:
Worker ID out of bounds 7 after deallocate
```

**Comment by cglagovichTT** [comment]
@tt-asaigal may have a fix for the deallocate out of bounds error. Some workers were inserting into the map at the same time as workers were looking into the map, which led to a race condition.

However, there still exists a segfault on main which is reproed by this same test. 

```
Fatal Python error: Segmentation fault

Thread 0x00007f7062ffd700 (most recent call first):
  File "/usr/lib/python3.8/threading.py", line 306 in wait
  File "/usr/lib/python3.8/threading.py", line 558 in wait
  File "/home/cglagovich/tt-metal/python_env/lib/python3.8/site-packages/tqdm/_monitor.py", line 60 in run
  File "/usr/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/usr/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007f73f1017740 (most recent call first):
  File "/home/cglagovich/tt-metal/ttnn/ttnn/core.py", line 37 in is_tensor_storage_on_device
  File "/home/cglagovich/tt-metal/ttnn/ttnn/operations/core.py", line 352 in to_torch
  File "/home/cglagovich/tt-metal/models/experimental/llama2_70b/tests/test_llama_perf_decode.py", line 130 in ru
```

**Comment by cglagovichTT** [comment]
@tt-asaigal resolved with #8963
